Implementation of fee adjustment

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#92 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added fee interface and corresponding types
- [x] I have added fee adjustment implementation
- [x] I have added unit tests
- [x] I have added property tests


# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ commented on src/Cardano/Wallet/API/V1/Types.hs:
```diff
@@ -1982,6 +1984,41 @@ instance BuildableSafeGen WalletAndTxHistory where
+instance ToSchema WalletSoftwareVersion where
+    declareNamedSchema _ =
+        pure $ NamedSchema (Just "WalletSoftwareVersion") $ mempty
+            & type_ .~ SwaggerObject
```
So... If we weren't defining any custom instances for `SoftwareVersion`.. What was `V1` used for ... ? Sounds like we could just remove it entirely without going for a newtype.

denisshevchenko replied:
I've tried it already, but it requires `ToSchema` for `SoftwareVersion`, it doesn't compile without it. But we cannot define orphan instances for `Core`-types, right?

KtorZ replied:
but where do the previous `ToSchema` instance came from... ?

denisshevchenko replied:
We have `instance ToSchema (V1 SoftwareVersion)` in `cardano-sl-chain` package.

KtorZ commented:
@denisshevchenko This conflicts with the PR on `Coin`.

denisshevchenko commented:
@KtorZ Fixed.

paweljakubas committed:
Implement first draft for fee adjustments

Remove pickUtxo argument and clean the code

KtorZ commented on src/Cardano/Wallet/CoinSelection/Fee.hs:
```diff
@@ -0,0 +1,351 @@
+    -- of outputs (incl. change) was bigger than the total input value which is
+    -- by definition, impossible; unless we messed up real hard.
+    collapse []  _ =
+        invariant "outputs are bigger than inputs" (Coin 0) (const True)
```
I believe it should be `const False` here, and let's replace `Coin 0` with `undefined` here ... this can never be reached.

paweljakubas replied:
indeed

KtorZ commented on src/Cardano/Wallet/CoinSelection/Fee.hs:
```diff
@@ -0,0 +1,351 @@
+    -> UTxO
+    -> m (Maybe ((TxIn, TxOut), UTxO))
+pickUtxoRandom _ utxo =
+    runMaybeT $ pickRandom utxo
```
I think this function / indirection is unnecessary here.

paweljakubas replied:
:+1:

KtorZ commented on src/Cardano/Wallet/CoinSelection/Fee.hs:
```diff
@@ -0,0 +1,351 @@
+
+
+coinToDouble :: Coin -> Double
+coinToDouble = fromIntegral . getCoin
```
We can probably simplify those above and not actually require `Double` here. We do know the underlying type for coins (`Word64`), so let's use that and arithmetic on integer (e.g. `div` or `quot` instead of `/`)

paweljakubas replied:
:+1:

KtorZ commented on src/Cardano/Wallet/CoinSelection/Fee.hs:
```diff
@@ -0,0 +1,351 @@
+    -- the goal fixed by 'estimatedFee'. So, the actualFee just can't be lower
+    -- than the goal.
+    if (   actualFee < estimatedFee
+        || (coinToDouble actualFee) / 2.0 > (coinToDouble estimatedFee)) then
```
same remark here about `Double` ... We know the underlying type, so let's favor this one.

paweljakubas replied:
:+1:

KtorZ commented on src/Cardano/Wallet/Fee.hs:
```diff
@@ -0,0 +1,48 @@
+-- Provides the API of Fee calculation.
+
+
+module Cardano.Wallet.Fee
```
No need for an extra module here, let's have that in `Cardano.Wallet.CoinSelection` :+1:

paweljakubas replied:
relocated `FeeOptions` and `FeeErrors` to `Cardano.Wallet.CoinSelection`

KtorZ commented on src/Cardano/Wallet/CoinSelection/Fee.hs:
```diff
@@ -0,0 +1,351 @@
+        removeDust = L.filter (> (dustThreshold opt))
+
+        reduceChangeOutputs
+            :: ([Coin] -> [Coin])
```
No need for the abstraction here, we can use `removeDust` directly from above.

paweljakubas replied:
:+1:

KtorZ commented on src/Cardano/Wallet/CoinSelection/Fee.hs:
```diff
@@ -0,0 +1,351 @@
+                remFee <- lift $ runMaybeT $ coverRemainingFee remainingFee utxo
+                case remFee of
+                    Nothing ->
+                        throwE CannotCoverFee
```
Maybe we can have the error a bit more explicit here :thinking: ? Saying that we cannot cover for fee because there's not enough inputs? And actually communicate about the missing amount?

paweljakubas replied:
added `missing amount` arg

KtorZ commented on src/Cardano/Wallet/CoinSelection/Fee.hs:
```diff
@@ -0,0 +1,351 @@
+    map (\a -> (feeForOut a, a)) outs
+  where
+    totalOut :: Coin
+    totalOut = (Coin . sum . map getCoin) outs
```
This seems like it could overflow :thinking: ...

paweljakubas replied:
outputs can add to more than upperBound of Coin. I used `Word64` here, and checking in `feeForOut` just below overflowing

KtorZ commented on src/Cardano/Wallet/Fee.hs:
```diff
@@ -0,0 +1,48 @@
+import GHC.Generics
+    ( Generic )
+
+newtype Fee = Fee { getFee :: Coin } deriving (Show, Eq)
```
I don't think we need the newtype here. Fee are just `Coin`

paweljakubas replied:
now no more `Fee` just `Coin`

KtorZ commented on cardano-wallet.cabal:
```diff
@@ -69,6 +69,7 @@ library
       Cardano.Wallet.Binary
       Cardano.Wallet.Binary.Packfile
       Cardano.Wallet.CoinSelection
+      Cardano.Wallet.CoinSelection.Fee
```
Let's completely remove this module and have it merged with `Cardano.Wallet.CoinSelection`

paweljakubas replied:
:+1:

KtorZ commented on src/Cardano/Wallet/CoinSelection/Random.hs:
```diff
@@ -11,6 +11,8 @@
 module Cardano.Wallet.CoinSelection.Random
     ( random
+    , pickRandom
+    , distance
```
Maybe distance is something we want to have in `Primitive.Types` :thinking: ?

paweljakubas replied:
done

KtorZ commented on src/Cardano/Wallet/CoinSelection/Fee.hs:
```diff
@@ -0,0 +1,336 @@
+computeFee
+    :: CoinSelection
+    -> Coin
+computeFee (CoinSelection inps outs chgs) =
```
Note that this whole algorithm here exists only because in the original implementation, we were in the generic realm, where there was no concrete instance for `Coin`. Here, we know that `Coin = Coin Word64` with `maxBound :: Coin` being much much less than `maxBound :: Word64`; We could in practice makes the computation `Word64` (or `Natural`) if we really really want the safety here, and at the end, convert it back to a Coin. 

Also, we can pretty much _safely_ assume (i.e. enforce with an invariant) that the resulting fee Coin is less than the max bound for Coin (otherwise that's a very very expensive transaction we have there!)

paweljakubas replied:
thanks for this comment

paweljakubas committed:
Adding tons of unit tests and revert to double conversion

paweljakubas commented on src/Cardano/Wallet/CoinSelection/Fee.hs:
```diff
@@ -0,0 +1,349 @@
+    totalOuts :: Word64
+    totalOuts = (sum . map getCoin) outs
+
+    feeForOut :: Coin -> Coin
```
for the simplicity and error plucking I am backed with double

paweljakubas commented on test/unit/Cardano/Wallet/CoinSelection/FeeSpec.hs:
```diff
@@ -0,0 +1,333 @@
+            , fExtraUtxo = [1]
+            , fFee = 5
+            , fDust = 0
+            }) (Left $ CannotCoverFee 2)
```
maybe should be `CannotCoverFee 1` as `fChngs` and `fExtraUtxo` gives 4 ... now because of MaybeT usage signalling about missing coins before random picking started

paweljakubas replied:
will be corrected in coming version

paweljakubas commented on test/unit/Cardano/Wallet/CoinSelection/FeeSpec.hs:
```diff
@@ -0,0 +1,333 @@
+            , fExtraUtxo = [2]
+            , fFee = 5
+            , fDust = 0
+            }) (Left $ OutOfBoundFee 0 5)
```
this one I do not understand. From `senderPaysFee` I have : 
```
coinSel' = CoinSelection {inputs = [(TxIn {inputId = Hash {getHash = "\170\223\131K\130@\DLE2\202\159"}, inputIx = 0},TxOut {address = Address {getAddress = "ADDR03"}, coin = Coin {getCoin = 10}}),(TxIn {inputId = Hash {getHash = "\219\CANf\153\159\&9\189\154\&9\133"}, inputIx = 0},TxOut {address = Address {getAddress = "ADDR01"}, coin = Coin {getCoin = 2}})], outputs = [TxOut {address = Address {getAddress = "ADDR01"}, coin = Coin {getCoin = 7}}], change = [Coin {getCoin = 5}]}
```
and after `computeFee coinSel'` I have Coin 0 ... The computation is : 
```
collapse : p>m : Coin {getCoin = 10} > Coin {getCoin = 7} ps:[Coin {getCoin = 2}] ms[Coin {getCoin = 5}]
collapse : p<m : Coin {getCoin = 3} < Coin {getCoin = 5} ps:[Coin {getCoin = 2}] ms[]
collapse : p==m : Coin {getCoin = 2} ps:[] ms[]
collapse : plus : []
```
I need to think about it - though it should be **Right**

paweljakubas replied:
addressed in new commit

paweljakubas committed:
Adding property tests

Finetuning tests and adding rest properties

Add estimate function pointers

Improving on properties

KtorZ committed:
style improvements & readability refactorings

KtorZ committed:
fix missing recursion after new inputs got selected to cover fee

KtorZ committed:
review properties and unit tests for fee adjustment

KtorZ committed:
add shrinkers & show error cases through Buildable instances

KtorZ committed:
fix LargestFirst to not return zero coin values as change
